### PR TITLE
Rule to conveniently use buildifier via Bazel

### PR DESCRIPTION
**Pull request by ash2k**
Updates #316.

### CODE REVIEW

File: buildifier/def.bzl [hunk 1]
```diff
@@ -0,0 +1,62 @@
+load("@bazel_skylib//:lib.bzl", "shell")
+
+def _buildifier_impl(ctx):
```
Comment (ash2k) [inline buildifier/def.bzl]: This is written looking at how Gazelle does it.

File: buildifier/runner.bash.template [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+BUILDIFIER_SHORT_PATH=@@BUILDIFIER_SHORT_PATH@@
+ARGS=@@ARGS@@
+
+buildifier_short_path=$(readlink "$BUILDIFIER_SHORT_PATH")
+cd "$BUILD_WORKSPACE_DIRECTORY"
```
Comment (ash2k) [inline buildifier/runner.bash.template]: This is using new Bazel 0.12 env variable. See https://blog.bazel.build/2018/04/11/bazel-0.12.html

File: buildifier/def.bzl [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    attrs = {
+        "verbose": attr.bool(
+            doc = "Print verbose information on standard error",
+        ),
+        "mode": attr.string(
+            default = "fix",
+            doc = "Formatting mode",
+            values = ["check", "diff", "fix"],
+        ),
+        "exclude_path_patterns": attr.string_list(
```
Comment (ash2k) [inline buildifier/def.bzl]: I'm not happy with this attribute name. Any suggestions?

Comment (laurentlb) [reply buildifier/def.bzl]: exclude_paths?
exclude_patterns?

Attributes can specify documentation (see 'doc' in https://docs.bazel.build/versions/master/skylark/lib/attr.html#string_list). Maybe mention here how it should look like.

File: buildifier/def.bzl [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            allow_empty = True,
+            default = [],
```
Comment (laurentlb) [inline buildifier/def.bzl]: No need to specify the default value (it's already the empty list).

**Comment by ash2k** [comment]
@laurentlb Thanks for review! I've addressed the comments. Also added "showlog" attribute.

**Comment by ash2k** [comment]
@laurentlb I'm a bazel newbie and I had these questions while working on this change:
- Is there a way to tell bazel to capture stdout from a `ctx.actions.run()` and put it into an output file? Creating a wrapper bash script just to do that is a bit annoying. Had this issue in another rule.
- Would be useful to be able to "attach" command line arguments to an executable action. Right now the only way to pass information is runfiles and/or a templated wrapper script (like in this PR). Is there a way to do this already? I've just copied what Gazelle does and was not able to find a better way. I.e. I'd like to do this:
```bzl
    return [DefaultInfo(
        files = depset([out_file]),
        runfiles = runfiles,
        executable = out_file,
        args = args,      # <------ THIS
    )]
```

File: WORKSPACE [hunk 5]
```diff
@@ -1,31 +1,38 @@
 workspace(name = "com_github_bazelbuild_buildtools")
 
-git_repository(
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
```
[the remark this answers is not in the document]
Comment (laurentlb) [reply WORKSPACE]: Thanks for this change!

**Comment by laurentlb** [comment]
Short answer is no, we don't provide a better way at the moment.

Capturing stdout a common limitation, we should do something for this.
Attaching command-line arguments is a good idea. I wonder if it should be built-in in Bazel, or if an implementation in a bzl file is enough.

Apparently tests are failing. Any idea what's going wrong?

**Comment by ash2k** [comment]
> Apparently tests are failing. Any idea what's going wrong?

I have rolled back rules_go and gazelle update and everything is fine now. I don't know what is wrong, something with protobuf generation if I'm understanding it correctly (I might be not). I'll leave it to someone who understands the problem to update them later.

**Comment by laurentlb** [comment]
Thanks!

**Comment by ash2k** [comment]
@laurentlb FYI I've filled https://github.com/bazelbuild/bazel/issues/5511

**Pull request merged**